### PR TITLE
fixed infinite forward in singleton rule nodes

### PR DESCRIPTION
**Pull request by YevhenBondarenko**
## Pull Request description

For now, Rule Engine that will run a singleton node is chosen based on partitionService.resolve(tenantId, "Main", ruleNodeId). When a message comes to a different Rule Engine, it should forward it to the one that is responsible for the rule node. But the message is forwarded to partitionService.resolve(tenantId, tbMsg.queueName, ruleNodeId), which is wrong. We should use the same queueName when deciding where to start rule node, and where to forward. 

This can be achieved by introducing queueName property to singleton nodes configuration.

The problem appears when using queues with different partitions count.

image-20231026-123156

https://thingsboard-portal.atlassian.net/browse/PROD-2527

## General checklist

- [x] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [x] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [x] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.  
- [x] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [x] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.
  
## Front-End feature checklist

- [ ] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [ ] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [ ] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [ ] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [x] If new dependency was added: the dependency tree is checked for conflicts.
- [x] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [x] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [x] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

File: application/src/main/java/org/thingsboard/server/actors/ruleChain/DefaultTbContext.java [hunk 1]
```diff
@@ -310,7 +310,12 @@ public void ack(TbMsg tbMsg) {
 
     @Override
     public boolean isLocalEntity(EntityId entityId) {
-        return mainCtx.resolve(ServiceType.TB_RULE_ENGINE, getTenantId(), entityId).isMyPartition();
+        return isLocalEntity(entityId, null);
+    }
+
+    @Override
+    public boolean isLocalEntity(EntityId entityId, String queueName) {
+        return mainCtx.resolve(ServiceType.TB_RULE_ENGINE, queueName, getTenantId(), entityId).isMyPartition();
```
[the remark this answers is not in the document]
Comment (ashvayka) [reply application/src/main/java/org/thingsboard/server/actors/ruleChain/DefaultTbContext.java:318]: I don't see where we set the queue name for any of the new/existing rule nodes. How is this going to happen?

[the remark this answers is not in the document]
Comment (YevhenBondarenko) [reply application/src/main/java/org/thingsboard/server/actors/ruleChain/DefaultTbContext.java:318]: Queue name we set on UI side. If the Queue name is null, then `Main` queue will be used.

[the remark this answers is not in the document]
Comment (YevhenBondarenko) [reply application/src/main/java/org/thingsboard/server/actors/ruleChain/DefaultTbContext.java:318]: > `DefaultTbContext` already has the `nodeCtx` which has the `RuleNode` which has the queueName now. So we can use it as is and why do we need to change the interface?

Ok, we use DefaultTbContext in RuleNodeActorMessageProcessor and can remove this method from the interface.

File: application/src/main/java/org/thingsboard/server/actors/ruleChain/RuleNodeActorMessageProcessor.java [hunk 2]
```diff
@@ -187,13 +187,13 @@ private boolean isMyNodePartition() {
     private boolean isMyNodePartition(RuleNode ruleNode) {
         return ruleNode == null || !ruleNode.isSingletonMode()
                 || systemContext.getDiscoveryService().isMonolith()
-                || defaultCtx.isLocalEntity(ruleNode.getId());
+                || defaultCtx.isLocalEntity(entityId);
```
Comment (ashvayka) [inline application/src/main/java/org/thingsboard/server/actors/ruleChain/RuleNodeActorMessageProcessor.java:190]: | defaultCtx.isLocalEntity(ruleNode.getId();

**Comment by YevhenBondarenko** [comment]
Fix available on: https://github.com/thingsboard/thingsboard/pull/9828

**Pull request closed without merging**